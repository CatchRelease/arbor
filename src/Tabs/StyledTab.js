import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import Box from '../Box';

const activeStyles = ({ active, theme }) => {
  const color = theme.colors.intent.brand.dark;

  return (
    active &&
    css`
      color: ${color};
      border-color: ${color};

      &:focus,
      &:hover {
        background: transparent;
      }
    `
  );
};

const StyledTab = styled(Box)`
  cursor: pointer;
  outline: none;

  &:last-child {
    margin-right: 0;
  }

  &:focus,
  &:hover {
    background: ${props => props.theme.colors.background.muted};
  }

  ${activeStyles};
`;

StyledTab.propTypes = {
  active: PropTypes.bool,
  tabIndex: PropTypes.string
};

StyledTab.defaultProps = {
  active: false,
  alignItems: 'center',
  borderBottom: '2px solid transparent',
  borderColor: 'border.default',
  color: 'text.muted',
  display: 'inline-block',
  fontSize: 'size4',
  justifyContent: 'center',
  mb: '-4px',
  mr: 'regular',
  overflow: 'hidden',
  py: 'regular',
  tabIndex: '0',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
};

export default StyledTab;
