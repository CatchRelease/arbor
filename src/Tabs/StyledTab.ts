import styled from '@emotion/styled';
import { css, Theme } from '@emotion/react';

import Box, { BoxProps } from '../Box';

const activeStyles = ({
  active,
  theme
}: {
  active?: boolean;
  theme: Theme;
}) => {
  const color = theme.colors.intent.brand.dark;

  return (
    active &&
    css`
      color: ${color};
      border-color: ${color};

      &:focus,
      &:hover {
        background: transparent;
        border-color: ${color};
      }
    `
  );
};

export type Props = BoxProps & {
  active?: boolean;
};

const StyledTab = styled(Box)<Props>`
  cursor: pointer;
  font-weight: ${(props) => props.theme.fontWeights.medium};
  line-height: ${(props) => props.theme.lineHeights.small};
  outline: none;

  &:last-child {
    margin-right: 0;
  }

  &:focus,
  &:hover {
    background: ${(props) => props.theme.colors.background.muted};
    border-color: ${(props) => props.theme.colors.border.muted};
  }

  ${activeStyles};
`;

StyledTab.defaultProps = {
  active: false,
  alignItems: 'center',
  borderBottom: '1px solid transparent',
  color: 'text.muted',
  display: 'inline-block',
  fontSize: 'size5',
  justifyContent: 'center',
  mb: '-4px',
  mr: 'large',
  overflow: 'hidden',
  py: 'regular',
  tabIndex: 0,
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
};

export default StyledTab;
