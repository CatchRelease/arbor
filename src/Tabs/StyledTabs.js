import Box from '../Box';

const StyledTabs = Box.withComponent('div');

StyledTabs.defaultProps = {
  ...Box.defaultProps,
  borderBottom: '2px solid transparent',
  borderColor: 'border.default',
  display: 'inline-block'
};

export default StyledTabs;
