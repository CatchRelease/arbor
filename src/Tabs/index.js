export { default as Tab } from './Tab';
export { default as TabContent } from './TabContent';
export { default as Tabs } from './Tabs';
