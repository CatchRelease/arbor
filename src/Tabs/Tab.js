import PropTypes from 'prop-types';

import StyledTab from './StyledTab';

const Tab = ({ children, ...props }) => (
  <StyledTab {...props}>{children}</StyledTab>
);

Tab.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,
  id: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  tabIndex: PropTypes.string
};

Tab.defaultProps = {
  active: false,
  onClick: null,
  tabIndex: '0'
};

export default Tab;
