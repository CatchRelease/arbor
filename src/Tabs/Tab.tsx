import { ReactElement, ReactNode } from 'react';

import StyledTab, { Props as StyledTabProps } from './StyledTab';

export type Props = StyledTabProps & {
  children: ReactElement<{ id?: string }>;
  id: string;
  onClick?: (callback: () => void) => void;
  title: ReactNode;
};

const Tab = ({ children, ...props }: Props) => (
  <StyledTab {...props}>{children}</StyledTab>
);

Tab.defaultProps = {
  active: false,
  tabIndex: 0
};

export default Tab;
