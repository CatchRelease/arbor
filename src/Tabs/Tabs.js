import React from 'react';
import PropTypes from 'prop-types';

import StyledTabs from './StyledTabs';
import Tab from './Tab';
import { ENTER_KEY, SPACEBAR } from '../constants';

const getTabContentId = ({ props: { children, id } }) =>
  children.props.id || `${id}-tab-content`;

class Tabs extends React.Component {
  constructor(props) {
    super(props);

    const { activeTabId, children } = this.props;

    this.state = {
      activeTabId: activeTabId || children[0].props.id
    };
  }

  get activeTab() {
    const { activeTabId } = this.state;
    const { children } = this.props;
    return children.find(({ props: { id } }) => id === activeTabId);
  }

  get activeTabContent() {
    const { activeTab } = this;

    return React.cloneElement(activeTab.props.children, {
      id: getTabContentId(activeTab)
    });
  }

  activateTab = tab => {
    const { id } = tab.props;

    this.setState({ activeTabId: id });
  };

  handleKeyPress = (key, tab, onClick) => {
    if (key === ENTER_KEY || key === SPACEBAR) {
      onClick();
    }
  };

  render() {
    const { activeTabId } = this.state;
    const { children, ...props } = this.props;

    return (
      <>
        <StyledTabs {...props}>
          {children.map(tab => {
            const { id, title, onClick: originalOnClick } = tab.props;
            const active = activeTabId === id;
            const tabContentId = getTabContentId(tab);
            const tabActivator = () => this.activateTab(tab);
            const onClick = originalOnClick
              ? () => originalOnClick(tabActivator)
              : tabActivator;

            return React.cloneElement(
              tab,
              {
                'aria-controls': tabContentId,
                'aria-selected': active ? 'true' : 'false',
                active,
                key: id,
                onClick,
                onKeyPress: ({ key }) => this.handleKeyPress(key, tab, onClick)
              },
              title
            );
          })}
        </StyledTabs>
        {this.activeTabContent}
      </>
    );
  }
}

Tabs.propTypes = {
  activeTabId: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(Tab),
    PropTypes.objectOf(Tab)
  ]).isRequired
};

Tabs.defaultProps = {
  activeTabId: undefined
};

export default Tabs;
