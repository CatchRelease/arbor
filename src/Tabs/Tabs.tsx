import {
  cloneElement,
  KeyboardEvent,
  ReactElement,
  ReactNode,
  useState
} from 'react';

import castArray from 'lodash/castArray';

import Box from '../Box';
import Flex from '../Flex';
import StyledTabs, { Props as StyledTabProps } from './StyledTabs';
import { Props as TabProps } from './Tab';
import { ENTER_KEY, SPACEBAR } from '../constants';

type Tab = ReactElement<TabProps>;

const getActiveTab = (
  activeTabId: string | undefined,
  children: Array<Tab | null>
) =>
  children.find((tab) => {
    if (!tab) {
      return false;
    }

    const {
      props: { id }
    } = tab;

    return id === activeTabId;
  });

const getTabContentId = (tab: Tab) => {
  if (!tab) {
    return undefined;
  }

  const {
    props: { children, id }
  } = tab;

  return children?.props.id || `${id}-tab-content`;
};

type Props = Omit<StyledTabProps, 'children'> & {
  activeTabId?: string;
  children: Tab | Array<Tab | null>;
  defaultTabId?: string;
  tabBarAside?: ReactNode;
};

const Tabs = ({
  activeTabId,
  children,
  defaultTabId,
  tabBarAside,
  ...props
}: Props) => {
  const childrenAsArray = castArray(children);

  const [state, setState] = useState(() =>
    activeTabId
      ? {}
      : {
          activeTabId:
            defaultTabId || childrenAsArray.find((tab) => !!tab)?.props.id
        }
  );

  const currentTabId = activeTabId || state.activeTabId;
  const activeTab = getActiveTab(currentTabId, childrenAsArray);

  if (!activeTab) {
    return null;
  }

  const activeTabContent = cloneElement(activeTab.props.children, {
    id: getTabContentId(activeTab)
  });

  const activateTab = (tab: Tab) => {
    setState({ activeTabId: tab.props.id });
  };

  const handleKeyPress = (key: string, onClick: () => void) => {
    if (key === ENTER_KEY || key === SPACEBAR) {
      onClick();
    }
  };

  const handleTabClick = (
    tab: Tab,
    originalOnClick?: (callback: () => void) => void
  ) => {
    const isControlled = !!activeTabId;
    const callback = isControlled ? () => {} : () => activateTab(tab);

    if (originalOnClick) {
      originalOnClick(callback);
    } else {
      callback();
    }
  };

  return (
    <>
      <StyledTabs {...props}>
        <Box>
          {childrenAsArray.map((tab) => {
            if (!tab) {
              return tab;
            }

            const { id, title, onClick: originalOnClick } = tab.props;
            const active = currentTabId === id;
            const tabContentId = getTabContentId(tab);
            const onClick = () => handleTabClick(tab, originalOnClick);

            return cloneElement(
              tab,
              {
                'aria-controls': tabContentId,
                'aria-selected': active ? 'true' : 'false',
                active,
                key: id,
                onClick,
                onKeyPress: ({ key }: KeyboardEvent<HTMLElement>) =>
                  handleKeyPress(key, onClick)
              },
              title
            );
          })}
        </Box>
        {tabBarAside && <Flex alignItems="center">{tabBarAside}</Flex>}
      </StyledTabs>
      {activeTabContent}
    </>
  );
};

Tabs.defaultProps = {
  activeTabId: undefined,
  defaultTabId: undefined,
  tabBarAside: undefined
};

export default Tabs;
