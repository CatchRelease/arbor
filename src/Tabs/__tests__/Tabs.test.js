import React from 'react';
import { shallow } from 'enzyme';

import Box from '../../Box';
import Tab from '../Tab';
import Tabs from '../Tabs';
import { ENTER_KEY, SPACEBAR } from '../../constants';

const render = (props = {}, tabProps = {}) =>
  shallow(
    <Tabs {...props}>
      <Tab id="tab-1" title="Tab 1" {...tabProps}>
        <Box>Tab 1 Content</Box>
      </Tab>
      <Tab id="tab-2" title="Tab 2">
        <Box>Tab 2 Content</Box>
      </Tab>
    </Tabs>
  );

describe('<Tabs />', () => {
  context('activeTabId is not passed', () => {
    it('renders the first tab as active', () => {
      const tabs = render();

      expect(tabs).toMatchSnapshot();
    });
  });

  context('activeTabId is passed', () => {
    it('renders the passed tab as active', () => {
      const tabs = render({ activeTabId: 'tab-2' });

      expect(tabs).toMatchSnapshot();
    });
  });

  describe('event handlers', () => {
    describe('clicking a tab', () => {
      context('tab has no onClick set', () => {
        it('activates the selected tab', () => {
          const tabs = render({ activeTabId: 'tab-1' });

          tabs.find('#tab-2').simulate('click');

          expect(tabs).toHaveState({ activeTabId: 'tab-2' });
        });
      });

      context('tab has an onClick set', () => {
        it('calls the passed onClick', () => {
          const onClick = jest.fn();
          const tabs = render({ activeTabId: 'tab-2' }, { onClick });
          tabs.find('#tab-1').simulate('click');
          expect(onClick).toHaveBeenCalledTimes(1);
        });

        it('passes a callback to activate the tab to the onClick', () => {
          const onClick = jest.fn(callback => {
            callback();
          });

          const tabs = render({ activeTabId: 'tab-2' }, { onClick });
          tabs.find('#tab-1').simulate('click');

          expect(tabs).toHaveState({ activeTabId: 'tab-1' });
        });
      });
    });

    const itHandlesKeypressesOnTabs = key => {
      context('no onClick is set on the tab', () => {
        it('activates the selected tab', () => {
          const tabs = render({ activeTabId: 'tab-1' });

          tabs.find('#tab-2').simulate('keyPress', { key });

          expect(tabs).toHaveState({ activeTabId: 'tab-2' });
        });
      });

      context('tab has an onClick set', () => {
        it('calls the onClick prop', () => {
          const onClick = jest.fn();
          const tabs = render({ activeTabId: 'tab-2' }, { onClick });
          tabs.find('#tab-1').simulate('keyPress', { key });
          expect(onClick).toHaveBeenCalledTimes(1);
        });

        it('passes a callback to activate the tab to the onClick', () => {
          const onClick = jest.fn(callback => {
            callback();
          });

          const tabs = render({ activeTabId: 'tab-2' }, { onClick });
          tabs.find('#tab-1').simulate('keyPress', { key });

          expect(tabs).toHaveState({ activeTabId: 'tab-1' });
        });
      });
    };

    describe('pressing enter on a tab', () => {
      itHandlesKeypressesOnTabs(ENTER_KEY);
    });

    describe('pressing space on a tab', () => {
      itHandlesKeypressesOnTabs(SPACEBAR);
    });
  });
});
