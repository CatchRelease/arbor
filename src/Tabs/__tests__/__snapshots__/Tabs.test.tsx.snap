// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> works with a single child tab 1`] = `
<Fragment>
  <Styled(div)
    borderBottom="1px solid transparent"
    borderColor="border.muted"
    display="inline-flex"
    width="100%"
  >
    <Styled(div)>
      <Tab
        active={true}
        aria-controls="tab-1-tab-content"
        aria-selected="true"
        id="tab-1"
        key="tab-1"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
        title="Tab 1"
      >
        Tab 1
      </Tab>
    </Styled(div)>
  </Styled(div)>
  <Styled(div)
    id="tab-1-tab-content"
  >
    Tab 1 Content
  </Styled(div)>
</Fragment>
`;
